File: qubo/frontend/src/types/game.ts
export interface GameMatch {
  match_id: string;
  user_id: string;
  winner: string | null;
  turns_played: number;
  match_date: string;
  game_level: number;
  waves_cleared: number;
  is_victory: boolean;
  score: number | null;
  enemies_defeated: number | null;
  compounds_discovered: number | null;
  highest_combo: number | null;
  leaderboard_points: number | null;
  completed_at: string | null;
}

export interface GameMatchHistory extends GameMatch {
  result: 'Win' | 'Loss';
}

export interface LevelOneLeaderboardEntry {
  rank: number;
  match_id: string;
  username: string;
  profile_picture_url: string | null;
  score: number;
  leaderboard_points: number;
  waves_cleared: number;
  enemies_defeated: number;
  compounds_discovered: number;
  highest_combo: number;
  grade: string;
  turns_played: number;
  completed_at: string | null;
  played_at: string;
}

export interface GameHistoryEvent {
  turn_number: number;
  effect_type?: string;
  effect_value?: number;
  corrupted_id?: string;
}

export interface UnityMatchStartedMessage {
  type: 'qubo:match-started';
}

export interface UnityMatchEventMessage extends GameHistoryEvent {
  type: 'qubo:match-event';
}

export interface UnityMatchCompletedMessage {
  type: 'qubo:match-completed';
  winner: string;
  turns_played: number;
  waves_cleared?: number;
  is_victory?: boolean;
  score?: number;
  enemies_defeated?: number;
  compounds_discovered?: number;
  highest_combo?: number;
}

export type UnityGameMessage =
  | UnityMatchStartedMessage
  | UnityMatchEventMessage
  | UnityMatchCompletedMessage;

File: qubo/backend/app/schemas/game.py
from typing import List, Optional
from uuid import UUID

from pydantic import BaseModel, Field


class MatchResponse(BaseModel):
    match_id: UUID
    user_id: UUID
    winner: Optional[str] = None
    turns_played: int
    match_date: str
    game_level: int = 1
    waves_cleared: int = 0
    is_victory: bool = False
    score: Optional[int] = None
    enemies_defeated: Optional[int] = None
    compounds_discovered: Optional[int] = None
    highest_combo: Optional[int] = None
    leaderboard_points: Optional[int] = None
    completed_at: Optional[str] = None


class MatchHistoryResponse(MatchResponse):
    result: str


class LeaderboardEntryResponse(BaseModel):
    rank: int
    match_id: UUID
    username: str
    profile_picture_url: Optional[str] = None
    score: int
    leaderboard_points: int
    waves_cleared: int
    enemies_defeated: int
    compounds_discovered: int
    highest_combo: int
    grade: str
    turns_played: int
    completed_at: Optional[str] = None
    played_at: str


class MatchHistoryEventCreate(BaseModel):
    turn_number: int = Field(ge=1, le=10_000)
    effect_type: Optional[str] = Field(default=None, max_length=100)
    effect_value: Optional[int] = Field(default=None, ge=-1_000_000, le=1_000_000)
    corrupted_id: Optional[UUID] = None


class MatchHistoryBatchCreate(BaseModel):
    events: List[MatchHistoryEventCreate] = Field(min_length=1, max_length=1_000)


class MatchHistoryBatchResponse(BaseModel):
    saved_count: int


class MatchCompleteRequest(BaseModel):
    winner: str = Field(min_length=1, max_length=100)
    turns_played: int = Field(ge=0, le=10_000)
    waves_cleared: int = Field(default=0, ge=0, le=100)
    is_victory: bool = False
    score: Optional[int] = Field(default=None, ge=0, le=10_000_000)
    enemies_defeated: Optional[int] = Field(default=None, ge=0, le=100_000)
    compounds_discovered: Optional[int] = Field(default=None, ge=0, le=100_000)
    highest_combo: Optional[int] = Field(default=None, ge=0, le=100_000)
